webapp/src/main/java/com/box/l10n/mojito/service/smartling/ThirdPartyTextUnitMatchingService.java

```diff
@@ -0,0 +1,173 @@
+                        stringDataToCheck.add(sourceStrings.getResponse().getData());
+                        offset += 500;
+                        sourceStrings = smartlingClient.getSourceStrings(projectId, fileUri, offset);
+                    }
+                }
+                stringDataToCheck.add(sourceStrings.getResponse().getData());
+
+                // pull all strings for a file then match
+                logger.debug("done pulling strings, starting matching");
+                matchSmartlingThirdPartyTextUnits(stringDataToCheck, fileUri);
+            }
+        }
+    }
+
+    private Boolean checkForFullMatchThirdPartyTextUnit(String hashcode, String mappingKey) {
+        ThirdPartyTextUnit existingThirdPartyTextUnitFullMatch = thirdPartyTextUnitRepository.findByThirdPartyTextUnitIdAndMappingKey(
+                hashcode,
+                mappingKey);
+        return existingThirdPartyTextUnitFullMatch != null;
+    }
+
+    private Boolean checkForPartialMatchThirdPartyTextUnit(ThirdPartyTextUnit thirdPartyTextUnit) {
+        ThirdPartyTextUnit existingThirdPartyTextUnitPartialMatch = thirdPartyTextUnitRepository.findByThirdPartyTextUnitId(
+                thirdPartyTextUnit.getThirdPartyTextUnitId());
+        if (existingThirdPartyTextUnitPartialMatch != null) {
+            thirdPartyTextUnitRepository.delete(existingThirdPartyTextUnitPartialMatch);
+            thirdPartyTextUnitRepository.flush();
+        }
+        return existingThirdPartyTextUnitPartialMatch != null;
+    }
+
+    private Boolean matchWithMojitoTextUnit(ThirdPartyTextUnit thirdPartyTextUnit, String name, String assetPath) {
+        TMTextUnit matchingMojitoTextUnit = tmTextUnitRepository.findFirstByNameAndAssetPath(name, assetPath);
+        if (matchingMojitoTextUnit == null) {
+            // If no matching mojito text unit, check to see if it is a plural
+            // smartling sends plurals without the plural suffix
+            matchingMojitoTextUnit = tmTextUnitRepository.findFirstByNameAndAssetPath(name + "_other", assetPath);
+        }
+        thirdPartyTextUnit.setTmTextUnit(matchingMojitoTextUnit);
+        return matchingMojitoTextUnit != null;
+    }
+
+
+    private void matchSmartlingThirdPartyTextUnits(List<StringData> stringDataList, String file) {
+        Instant start = Instant.now();
+        AtomicInteger fullMatch = new AtomicInteger(0);
+        AtomicInteger partialMatch = new AtomicInteger(0);
+        AtomicInteger newThirdPartyTextUnit = new AtomicInteger(0);
+        AtomicInteger newTPTWithoutMatchingMojitoTU = new AtomicInteger(0);
+        AtomicInteger missingVariant = new AtomicInteger(0);
+        String delimiter = "#@#";
+        // partition the list of StringData into groups of 500
+        List<List<StringData>> batchedStringData = Lists.partition(stringDataList, 500);
+
+        batchedStringData.forEach(
+                stringDataBatch -> {
```

Comment:
I just had a quick look and I need to do more digging... but here is my first feedback to start the discussion

can we split in 2 steps: "mapping" then "saving"? The batching logic here useful only when saving. Having all done at once is.

Also can you introduce functions to make the code clearer. At first glance, it uses "streams" but it is pretty much imperative loops so in that case it should be just standard "for loops". Or it could be rewritten fully in a more functional style but that'd probably start with `getSmartlingThirdPartyTextUnits` as well